### PR TITLE
Build NFT auction capability for card-store

### PR DESCRIPTION
**Pull request by simpletrontdip**
Closes https://github.com/agoric/agoric-sdk/issues/3882 (I am approved to work on that ticket)

- Modify Zoe's auction contract
  - Support **first-price** option
  - Correct set wakeup after the first bid
  - Add tests to verify new logic while old ones still pass
- Add **auctionItems** contract, to auction a list of items, with utilities for checking item state
- Update app code to deploy new contract and update UI as requirements

Please let me know if anything needs to be changed. ^^

Screen Shot 2021-10-28 at 23 00 59
Screen Shot 2021-10-28 at 23 06 22

**Update**: add close btn
Screen Shot 2021-11-07 at 11 23 42

### CODE REVIEW

**Review by katelynsills — CHANGES_REQUESTED** [review]
Hi @simpletrontdip, this looks excellent! It looks like you really understood what you were doing. 

This review does not include the UI - @samsiegart is looking at that - but I did review all of the contracts, deploy scripts, and tests. Just a few changes requested.

After those changes are done, let's take the contract files and move them to `agoric-sdk/packages/zoe/src/contracts` to replace the auction code there. Then this PR can import those files.

**Comment by katelynsills** [comment]
Hi @simpletrontdip, @samsiegart noticed that getting the card details actually launches the auction contract. Can you change this so that an auction contract instance is only created when someone submits a bid? Launching a contract instance is expensive, so we'd like to only create contract instances when needed.

File: contract/src/auctionItems.js
Comment (simpletrontdip) [reply contract/src/auctionItems.js:142]: Nice catch, lemme think about it.

File: contract/test/auction/test-auction.js
Comment (simpletrontdip) [reply contract/test/auction/test-auction.js:296]: Hi @katelynsills, this is actually not a test assertion, it is the global `assert` to eliminate the case `aliceInvitation` is null, correctly handle ts-check

File: contract/src/auctionItems.js
Comment (simpletrontdip) [reply contract/src/auctionItems.js:184]: Not sure how should I update this, all my approaches seem harder to read. 🤔

**Comment by simpletrontdip** [comment]
@katelynsills other than delete file comments (which will be resolved if we create another PR) and `conditional-await` (I'm not sure how to make it readable with Promise) then all the others are addressed.
Once you approve this, I will copy the contract file and create another PR for `agroic-sdk`, and rebase this one too.

File: ui/src/components/CardDetailModal.jsx
Comment (samsiegart) [inline ui/src/components/CardDetailModal.jsx:71]: This is not likely a big deal, but this might be a race condition. If playerName changes before this completes, it's possible the details could be updated with the old playerName. One way around this is to use a cleanup function to cancel if playerName changes https://maxrozen.com/race-conditions-fetching-data-react-with-useeffect#useeffect-clean-up-function-with-boolean-flag

**Comment by samsiegart** [comment]
LGTM! A couple small nitpicks: it seems like the card is "clickable" inside the modal even though it has no effect, maybe it shouldn't be. Also, the modal could use a close button.

File: contract/src/auctionItems.js
Comment (simpletrontdip) [reply contract/src/auctionItems.js:184]: Is this OK?
```js
const getOrCreateSessionP = (itemKey) => {
  const session = sellerSessions[itemKey];

  if (session) {
    return Promise.resolve(session);
  }

  return startAuctioningItem(itemKey).then((value) => {
    sellerSessions[itemKey] = value;
    return value;
  });
};

// later
session = await getOrCreateSessionP(itemKey)
```

**Review by katelynsills — APPROVED** [review]
This is only a review of the contract and deploy script code, but everything looks great to me. Thank you so much for making those changes. 

@samsiegart is out today, but will be taking a look at the UI. Please wait to merge until we get his approval on the UI. 

Really well done, thank you!

**Review by samsiegart — APPROVED** [review]
UI LGTM, thank you!

File: ui/src/auction.js
Comment (katelynsills) [inline ui/src/auction.js:39]: Hi @simpletrontdip, sorry, I missed this earlier, but does using the `amounts` here work to make an offer in the wallet?

Comment (simpletrontdip) [reply ui/src/auction.js:39]: Hi, as in the `complieProposal`, we need the `purseName`, so the original amount may not work
https://github.com/Agoric/agoric-sdk/blob/master/packages/dapp-svelte-wallet/api/src/lib-wallet.js#L849-L879
I think, I tried this before

**Pull request merged**
